inst/include/dust/cuda/dust_device.hpp

```diff
@@ -0,0 +1,797 @@
+      n_state_full_ = p.size();
+      n_state_ = n_state_full_;
+    }
+
+    initialise_device_memory(pars[0].shared);
+    set_device_shared(pars);
+    set_state_from_pars(pars);
+
+    // Set GPU RNG from a seed; primary reason for this construction
+    // is to expand out seed correctly (e.g., it might be a 4-element
+    // vector on first creation).
+    dust::random::prng<rng_state_type> rng(n_particles_total_ + 1, seed);
+    set_rng_state(rng.export_state());
+
+    set_cuda_launch();
+
+    select_needed_ = true;
+
+#ifdef DUST_USING_CUDA_PROFILER
+    cuda_profiler_start(device_config_);
+#endif
+  }
+
+  // This only gets called on construction; the size of these never
+  // changes.  Could be merged with the above really.
+  void initialise_device_memory(typename dust::shared_ptr<T> s) {
+    const size_t n_pars = n_pars_effective();
+    const size_t n_internal_int = dust::cuda::device_internal_int_size<T>(s);
+    const size_t n_internal_real = dust::cuda::device_internal_real_size<T>(s);
+    const size_t n_shared_int = dust::cuda::device_shared_int_size<T>(s);
+    const size_t n_shared_real = dust::cuda::device_shared_real_size<T>(s);
+    device_state_.initialise(n_particles_total_, n_state_full_, n_pars,
+                             n_internal_int, n_internal_real,
+                             n_shared_int, n_shared_real);
+  }
+
+  void set_device_shared(const std::vector<pars_type>& pars) {
+    size_t n = n_particles() == 0 ? 0 : n_state_full();
+    std::vector<dust::Particle<T>> p;
+    for (size_t i = 0; i < n_pars_effective(); ++i) {
+      p.push_back(dust::Particle<T>(pars[i], step_));
+      if (n > 0 && p.back().size() != n) {
+        std::stringstream msg;
+        msg << "'pars' created inconsistent state size: " <<
+          "expected length " << n << " but parameter set " << i + 1 <<
+          " created length " << p.back().size();
+        throw std::invalid_argument(msg.str());
+      }
+      n = p.back().size(); // ensures all particles have same size
+    }
+
+    const size_t n_shared_int = device_state_.n_shared_int;
+    const size_t n_shared_real = device_state_.n_shared_real;
+    std::vector<int> shared_int(n_shared_int * n_pars_effective());
+    std::vector<real_type> shared_real(n_shared_real * n_pars_effective());
+    for (size_t i = 0; i < pars.size(); ++i) {
+      int * dest_int = shared_int.data() + n_shared_int * i;
+      real_type * dest_real = shared_real.data() + n_shared_real * i;
+      dust::cuda::device_shared_copy<T>(pars[i].shared, dest_int, dest_real);
+    }
+    device_state_.shared_int.set_array(shared_int);
+    device_state_.shared_real.set_array(shared_real);
+  }
+
+  // TODO: This update function is wildly inefficient; we should
+  // probably support things like "copy one state to all the particles
+  // of that parameter index", possibly as a kernel.
+  void set_state_from_pars(const std::vector<pars_type>& pars) {
```

Comment:
It can be yes, but I am not sure we cope correctly with that yet and we do not test it anywhere - I think this is wrong in CPU code too https://github.com/mrc-ide/dust/issues/310 (added to comment too)